canopen_ros2_control/src/cia402_system.cpp

```diff
@@ -0,0 +1,215 @@
+                                                                           &motor_data_[node_id].velocity_mode.resp));
+
+      // set cyclic velocity mode
+      command_interfaces.emplace_back(hardware_interface::CommandInterface(info_.joints[i].name, "cyclic_velocity_mode_cmd",
+                                                                           &motor_data_[node_id].cyclic_velocity_mode.ons_cmd));
+      command_interfaces.emplace_back(hardware_interface::CommandInterface(info_.joints[i].name, "cyclic_velocity_mode_resp",
+                                                                           &motor_data_[node_id].cyclic_velocity_mode.resp));
+      // set cyclic position mode
+      command_interfaces.emplace_back(hardware_interface::CommandInterface(info_.joints[i].name, "cyclic_position_mode_cmd",
+                                                                           &motor_data_[node_id].cyclic_position_mode.ons_cmd));
+      command_interfaces.emplace_back(hardware_interface::CommandInterface(info_.joints[i].name, "cyclic_position_mode_resp",
+                                                                           &motor_data_[node_id].cyclic_position_mode.resp));
+
+  }
+
+  return command_interfaces;
+}
+
+hardware_interface::CallbackReturn Cia402System::on_activate(
+  const rclcpp_lifecycle::State & previous_state)
+{
+
+  return CanopenSystem::on_activate(previous_state);
+
+}
+
+hardware_interface::CallbackReturn Cia402System::on_deactivate(
+  const rclcpp_lifecycle::State & previous_state)
+{
+    return CanopenSystem::on_deactivate(previous_state);
+}
+
+hardware_interface::return_type Cia402System::read(
+  const rclcpp::Time & time, const rclcpp::Duration & period)
+{
+  // TODO(anyone): read robot states
+
+   auto ret_val = CanopenSystem::read(time, period);
+
+   auto drivers = device_container_->get_registered_drivers();
+
+   for(auto it = canopen_data_.begin(); it != canopen_data_.end(); ++it){
+       auto motion_controller_driver = std::static_pointer_cast<ros2_canopen::Cia402Driver>(drivers[it->first]);
+        // get position
+       motor_data_[it->first].actual_position = motion_controller_driver->get_position();
+       // get speed
+       motor_data_[it->first].actual_speed = motion_controller_driver->get_speed();
+   }
+
+    return ret_val;
+}
+
+hardware_interface::return_type Cia402System::write(
+  const rclcpp::Time & time, const rclcpp::Duration & period)
+{
+    auto ret_val = CanopenSystem::write(time, period);
+
+    auto drivers = device_container_->get_registered_drivers();
+
+    for(auto it = canopen_data_.begin(); it != canopen_data_.end(); ++it){
+        //TODO(livanov93): check casting
+        auto motion_controller_driver = std::static_pointer_cast<ros2_canopen::Cia402Driver>(drivers[it->first]);
+        //TODO(livanov93) check the mode and set target accordingly
+         motion_controller_driver->set_target(motor_data_[it->first].target.position_value);
```

Comment:
Another option would be this.

```suggestion
        auto node_canopen_402_driver = 
          std::static_pointer_cast<ros2_canopen::node_interfaces::NodeCanopen402Driver<rclcpp::Node>>(drivers[it->first]->get_node_canopen_driver_interface());
        // get position
       motor_data_[it->first].actual_position = node_canopen_402_driver->get_position();
       // get speed
       motor_data_[it->first].actual_speed = node_canopen_402_driver->get_speed();
```